crates/ruff_python_semantic/src/red_knot/types/infer.rs

```diff
@@ -0,0 +1,980 @@
+        let file_scope_id = scope.file_id(db);
+        let file = scope.file(db);
+        let children_scopes = index.child_scopes(file_scope_id);
+        let symbol_table = index.symbol_table(file_scope_id);
+
+        Self {
+            index,
+            file_scope_id,
+            file_id: file,
+            scope,
+            symbol_table,
+
+            db,
+            types: TypeInference::default(),
+            definition_tys: FxHashMap::default(),
+            children_scopes,
+        }
+    }
+
+    /// Lowers `module` to types.
+    pub(super) fn lower_module(&mut self, module: &ast::ModModule) {
+        self.lower_body(&module.body);
+    }
+
+    pub(super) fn lower_class_type_params(&mut self, class: &ast::StmtClassDef) {
+        if let Some(type_params) = class.type_params.as_deref() {
+            self.lower_type_parameters(type_params);
+        }
+    }
+
+    pub(super) fn lower_class_body(&mut self, class: &ast::StmtClassDef) {
+        self.lower_body(&class.body);
+    }
+
+    pub(super) fn lower_function_type_params(&mut self, function: &ast::StmtFunctionDef) {
+        if let Some(type_params) = function.type_params.as_deref() {
+            self.lower_type_parameters(type_params);
+        }
+    }
+
+    pub(super) fn lower_function_body(&mut self, function: &ast::StmtFunctionDef) {
+        self.lower_body(&function.body);
+    }
+
+    fn lower_body(&mut self, suite: &[ast::Stmt]) {
+        for statement in suite {
+            self.lower_statement(statement);
+        }
+    }
+
+    fn lower_statement(&mut self, statement: &ast::Stmt) {
+        match statement {
+            ast::Stmt::FunctionDef(function) => self.lower_function_definition_statement(function),
+            ast::Stmt::ClassDef(class) => self.lower_class_definition_statement(class),
+            ast::Stmt::Expr(ast::StmtExpr { range: _, value }) => {
+                self.lower_expression(value);
+            }
+            ast::Stmt::If(if_statement) => self.lower_if_statement(if_statement),
+            ast::Stmt::Assign(assign) => self.lower_assignment_statement(assign),
+            ast::Stmt::AnnAssign(assign) => self.lower_annotated_assignment_statement(assign),
+            ast::Stmt::For(for_statement) => self.lower_for_statement(for_statement),
+            ast::Stmt::Import(import) => self.lower_import_statement(import),
+            ast::Stmt::ImportFrom(import) => self.lower_import_from_statement(import),
+            ast::Stmt::Break(_) | ast::Stmt::Continue(_) | ast::Stmt::Pass(_) => {
+                // No-op
+            }
+            _ => {}
+        }
+    }
+
+    fn lower_function_definition_statement(&mut self, function: &ast::StmtFunctionDef) {
+        let ast::StmtFunctionDef {
+            range: _,
+            is_async: _,
+            name,
+            type_params: _,
+            parameters: _,
+            returns,
+            body: _,
+            decorator_list,
```

Comment:
The benefit of using `: _` over `..` is that it forces you to handle every field explicitly. Either by using it in the code below or explicitly ignoring it by adding a `: _`. 

The benefit of `..` is that you don't have to do that. 

My rule of thumb is that if some logic will likely need updating when a new AST field is added, use `: _`. If the logic concerns only a very specific aspect of the node, use `..`. 

For type inference I think it's important that Rust forces us to think about if the type inference logic needs updating when adding a new AST field, because most likely it does. That's why I use `: _`